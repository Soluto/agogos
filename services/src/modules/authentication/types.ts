import { ClientOptions } from 'jwks-rsa';

export interface IssuerConfig {
  authority: string;
  jwksConfig?: Partial<ClientOptions>;
  audience?: string | string[];
  authenticatedPaths: string[];
  description?: string;
}

export interface AuthenticationConfig {
  jwt?: Record<string, IssuerConfig>;
  anonymous?: {
    publicPaths: string[];
    rejectAuthorizationHeader?: boolean;
  };
}
