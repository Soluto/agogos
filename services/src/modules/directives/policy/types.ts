import {PolicyArgsObject, PolicyAttachments} from '../../resource-repository/types';

export type Policy = {
    namespace: string;
    name: string;
    args?: PolicyArgsObject;
};

// args here contain the final values after param injection
export type PolicyExecutionContext = {
    namespace: string;
    name: string;
    policyAttachments: PolicyAttachments;
    args?: PolicyArgsObject;
    query?: QueryResults;
};

export type QueryResults = {
    [name: string]: any;
};

export type PolicyExecutionResult = {
    done: boolean;
    allow?: boolean;
    query?: {
        type: string;
        code: string;
    };
};

export type GraphQLArguments = {
    [name: string]: any;
};
