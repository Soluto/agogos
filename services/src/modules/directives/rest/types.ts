export interface KeyValue {
  key: string;
  value: string;
}

export interface RestParams {
  url: string;
  method?: string;
  body?: string;
  bodyArg?: string;
  query?: KeyValue[];
  headers?: KeyValue[];
  timeoutMs?: number;
}
