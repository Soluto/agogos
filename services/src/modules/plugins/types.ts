import { BaseSchema } from '../base-schema';
import { ResourceGroup } from '../resource-repository';

export type ValueOrPromise<T> = T | Promise<T>;

export interface StitchPlugin {
  name: string;

  addArgumentInjectionGlobals?(): ValueOrPromise<Record<string, unknown>>;

  transformResourceGroup?(resourceGroup: ResourceGroup): ValueOrPromise<ResourceGroup>;

  transformBaseSchema?(baseSchema: BaseSchema): ValueOrPromise<BaseSchema>;
}
