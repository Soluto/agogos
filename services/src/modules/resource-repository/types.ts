import { RemoteSchema } from '../directives/gql';
import { Policy, LoadedPolicy } from '../directives/policy/types';

export type PolicyAttachments = Record<string, LoadedPolicy>;
export type PolicyQueryVariables = Record<string, unknown>;
export type PolicyArgsObject = Record<string, unknown>;

export interface ResourceGroup {
  schemas: Schema[];
  upstreams: Upstream[];
  upstreamClientCredentials: UpstreamClientCredentials[];
  policies: PolicyDefinition[];
  // policyAttachments are compiled from the Rego code in opa policies, they are not directly modified by users
  policyAttachments?: PolicyAttachments;
  basePolicy?: Policy;
  remoteSchemas?: RemoteSchema[];
}

export interface Resource {
  metadata: ResourceMetadata;
}

export interface ResourceMetadata {
  namespace: string;
  name: string;
}

export interface Schema extends Resource {
  schema: string;
}

export interface Upstream extends Resource {
  host: string;
  auth: {
    type: AuthType;
    activeDirectory: {
      authority: string;
      resource: string;
    };
  };
}

export interface UpstreamClientCredentials extends Resource {
  authType: AuthType;
  activeDirectory: {
    authority: string;
    clientId: string;
    clientSecret: string;
  };
}

export interface PolicyDefinition extends Resource {
  type: PolicyType;
  code: string;
  args?: PolicyArgsObject;
  query?: PolicyQuery;
  shouldOverrideBasePolicy?: boolean;
}

export interface PolicyQuery {
  gql: string;
  variables?: PolicyQueryVariables;
}

export interface FetchLatestResult {
  isNew: boolean;
  resourceGroup: ResourceGroup;
}

export type UpdateOptions = { registry: boolean };

export interface IResourceRepository {
  fetchLatest(): Promise<FetchLatestResult>;
  update(rg: ResourceGroup, options?: UpdateOptions): Promise<void>;
  writePolicyAttachment(filename: string, content: Buffer): Promise<void>;
}

enum AuthType {
  ActiveDirectory = 'ActiveDirectory',
}

export enum PolicyType {
  opa = 'opa',
}
