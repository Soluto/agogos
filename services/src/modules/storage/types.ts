export interface Storage {
  fileStats(filePath: string): Promise<FileStats>;
  writeFile(filePath: string, content: string | Buffer): Promise<void>;
  listFiles(folderPath: string): Promise<listFilesItem[]>;

  readFile(filePath: string): Promise<{ content: Buffer }>;
  readFile(filePath: string, options: { asString?: false }): Promise<{ content: Buffer }>;
  readFile(filePath: string, options: { asString?: true }): Promise<{ content: string }>;
  readFile(filePath: string, options?: { asString?: boolean }): Promise<{ content: Buffer | string }>;
}

export type FileStats = {
  lastModified: Date;
};

export type listFilesItem = {
  filePath: string;
  lastModified: Date;
};
