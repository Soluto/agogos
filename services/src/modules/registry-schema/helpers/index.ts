import proceedPolicy from './policy-helper';

export { proceedPolicy };
