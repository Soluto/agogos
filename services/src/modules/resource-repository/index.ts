export * from './types';

export { CompositeResourceRepository } from './composite';
export { S3ResourceRepository } from './s3';
export { FileSystemResourceRepository } from './fs';
export { pollForUpdates } from './stream';
export { applyResourceGroupUpdates } from './updates';
