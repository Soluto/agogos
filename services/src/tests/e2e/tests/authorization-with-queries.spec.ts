import { GraphQLClient } from 'graphql-request';
import { createSchemaMutation } from '../../helpers/authz-schema';
import { Policy, PolicyType, Schema } from '../../../modules/resource-repository/types';
import { sleep } from '../../helpers/utility';

const policies: Policy[] = [
  {
    metadata: {
      name: 'alwaysDenied',
      namespace: 'namespace',
    },
    type: PolicyType.opa,
    code: `
      default allow = false
        `,
  },
  {
    metadata: {
      name: 'isSenior',
      namespace: 'namespace',
    },
    type: PolicyType.opa,
    code: `
      default allow = false
      allow {
        input.args.hireDate < 2015
      }
    `,
    args: {
      hireDate: 'Int',
    },
  },
  {
    metadata: {
      name: 'notClassified',
      namespace: 'namespace',
    },
    type: PolicyType.opa,
    code: `
      default allow = false
      allow {
        input.query.classifiedDepartments[_].id != input.args.departmentId;
        input.query.policy.namespace___isSenior.allow
      }
    `,
    args: {
      departmentId: 'String',
      hireDate: 'Int',
    },
    query: {
      gql: `
        query($hireDate: Int!) {
          classifiedDepartments {
            id
          },
          policy {
            namespace___isSenior(hireDate: $hireDate) {
              allow
            }
          }
        }
      `,
      variables: {
        hireDate: '{hireDate}',
      },
    },
  },
];

const schema: Schema = {
  metadata: {
    name: 'EmployeeSchema',
    namespace: 'namespace',
  },
  schema: `
    type Department {
      id: String
      name: String
    }
    type Employee {
      id: String
      name: String
      hireDate: Int
      department: Department!
      address: String @policy(
        namespace: "namespace",
        name: "notClassified",
        args: {
          departmentId: "{source.department.id}",
          hireDate: "{source.hireDate}"
        }
      )
    }

    type Query {
      allowedEmployee: Employee! @stub(value: {
        id: "1"
        name: "John Smith"
        address: "Tel Aviv"
        hireDate: 2010
        department: {
          id: "D1"
          name: "Sales"
        }
      })
      deniedEmployee1: Employee! @stub(value: {
        id: "2"
        name: "Mark Zuckerberg",
        department: {
          id: "D1000"
          name: "VIP"
        }
        hireDate: 2010
        address: "Facebook HQ"
      })
      deniedEmployee2: Employee! @stub(value: {
        id: "2"
        name: "Tom Baker",
        department: {
          id: "D2"
          name: "VIP"
        }
        hireDate: 2019
        address: "Atlanta"
      })
      classifiedDepartments: [Department!]! @stub(value: [{
        id: "D1000"
        name: "VIP"
      }]) @policy(namespace: "namespace", name: "alwaysDenied")
    }
  `,
};

const createPolicyMutation = `
  mutation CreatePolicy($policies: [PolicyInput!]!) {
    updatePolicies(input: $policies) {
      success
    }
  }`;

const employeeQuery = (query: string) => `
  query {
    ${query} {
      id
      name
      address
    }
  }
`;

interface CreatePolicyMutationResponse {
  updatePolicies: {
    success: boolean;
  };
}

interface UpdateSchemasMutationResponse {
  updateSchemas: {
    success: boolean;
  };
}

interface AllowedEmployeeQueryResponse {
  allowedEmployee: {
    id: string;
    name: string;
    address: string;
  };
}

describe('Authorization with queries', () => {
  let gatewayClient: GraphQLClient;
  let registryClient: GraphQLClient;

  beforeAll(() => {
    gatewayClient = new GraphQLClient('http://localhost:8080/graphql');
    registryClient = new GraphQLClient('http://localhost:8090/graphql');
  });

  test('Setup policies', async () => {
    const policyResponse: CreatePolicyMutationResponse = await registryClient.request(createPolicyMutation, {
      policies,
    });
    expect(policyResponse.updatePolicies.success).toBeTruthy();

    const schemaResponse: UpdateSchemasMutationResponse = await registryClient.request(createSchemaMutation, {
      schema,
    });
    expect(schemaResponse.updateSchemas.success).toBeTruthy();

    // Wait for gateway to update before next tests
    await sleep(500);
  });

  test('Query allowed employee', async () => {
    // TODO: check classified fails
    try {
      await gatewayClient.request(`query {
              classifiedDepartments {
                  id
                  name
              }
          }`);
    } catch (e) {
      const response = e.response;
      expect(response).toMatchSnapshot({
        extensions: expect.any(Object),
      });
    }

    const response: AllowedEmployeeQueryResponse = await gatewayClient.request(employeeQuery('allowedEmployee'));
    expect(response.allowedEmployee).toMatchSnapshot();
  });

  test('Query denied employee 1', async () => {
    try {
      await gatewayClient.request(employeeQuery('deniedEmployee1'));
      expect(true).toBeFalsy();
    } catch (e) {
      const response = e.response;
      expect(response).toMatchSnapshot({
        extensions: expect.any(Object),
      });
    }
  });

  test('Query denied employee 2', async () => {
    try {
      await gatewayClient.request(employeeQuery('deniedEmployee2'));
      expect(true).toBeFalsy();
    } catch (e) {
      const response = e.response;
      expect(response).toMatchSnapshot({
        extensions: expect.any(Object),
      });
    }
  });
});
