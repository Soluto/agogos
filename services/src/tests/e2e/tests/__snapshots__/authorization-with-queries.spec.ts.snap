// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authorization with queries Query allowed employee 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": Object {
          "stacktrace": Array [
            "Error: Unauthorized by policy alwaysDenied in namespace namespace",
            "    at Function.validatePolicy (/service/dist/modules/directives/policy/policy-executor.js:28:19)",
            "    at async field.resolve (/service/dist/modules/directives/policy/policy.js:14:17)",
          ],
        },
      },
      "locations": Array [
        Object {
          "column": 15,
          "line": 2,
        },
      ],
      "message": "Unauthorized by policy alwaysDenied in namespace namespace",
      "path": Array [
        "classifiedDepartments",
      ],
    },
  ],
  "extensions": Any<Object>,
  "status": 200,
}
`;

exports[`Authorization with queries Query allowed employee 2`] = `
Object {
  "address": "Tel Aviv",
  "id": "1",
  "name": "John Smith",
}
`;

exports[`Authorization with queries Query denied employee 1 1`] = `
Object {
  "data": Object {
    "deniedEmployee1": Object {
      "address": null,
      "id": "2",
      "name": "Mark Zuckerberg",
    },
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": Object {
          "stacktrace": Array [
            "Error: Unauthorized by policy notClassified in namespace namespace",
            "    at Function.validatePolicy (/service/dist/modules/directives/policy/policy-executor.js:28:19)",
            "    at async field.resolve (/service/dist/modules/directives/policy/policy.js:14:17)",
          ],
        },
      },
      "locations": Array [
        Object {
          "column": 7,
          "line": 6,
        },
      ],
      "message": "Unauthorized by policy notClassified in namespace namespace",
      "path": Array [
        "deniedEmployee1",
        "address",
      ],
    },
  ],
  "extensions": Any<Object>,
  "status": 200,
}
`;

exports[`Authorization with queries Query denied employee 2 1`] = `
Object {
  "data": Object {
    "deniedEmployee2": Object {
      "address": null,
      "id": "2",
      "name": "Tom Baker",
    },
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": Object {
          "stacktrace": Array [
            "Error: Unauthorized by policy notClassified in namespace namespace",
            "    at Function.validatePolicy (/service/dist/modules/directives/policy/policy-executor.js:28:19)",
            "    at async field.resolve (/service/dist/modules/directives/policy/policy.js:14:17)",
          ],
        },
      },
      "locations": Array [
        Object {
          "column": 7,
          "line": 6,
        },
      ],
      "message": "Unauthorized by policy notClassified in namespace namespace",
      "path": Array [
        "deniedEmployee2",
        "address",
      ],
    },
  ],
  "extensions": Any<Object>,
  "status": 200,
}
`;
